Refactor to enhance code re-use and prepare for HTTP/3 support

gseddon opened the pull request:
Hi,
A bigger change here but I hope it is valuable! We want to contribute HTTP/3 support to Oha. This is an initial CR which refactors the `client.rs` to use common functions for http1 and http2 load generation. 
It also introduces a HttpWorkType enum which is matched on to choose which functionality to execute, rather than the previous is_http2() function. 

This lays the groundwork for adding a H3 work type in a following PR, and a http3 load generation common function.
This will be feature gated due to its experimental nature. 
This has been split off into a separate PR to make them easier to understand and review individually. It's nice to be able to have a refactor PR which doesn't need to change any tests - it is internal structural changes only. 

This PR has had an initial review here https://github.com/gseddon/oha/pull/1
and here is the HTTP3 branch, will make that PR once this one is merged https://github.com/gseddon/oha/pull/2

hatoo commented:
Hi, `HttpWorkType` looks good.
But please leave load generation functions original.
Your changes are bad for performance, and it's not working on my machine.

```
cargo run -- -z 10s http://127.0.0.1:3000
# stop after 50 requests ... 
```

gseddon commented:
Hi, I fixed the performance bug for HTTP1 with 'deadline' enabled. Good catch sorry. As far as performance impact, these changes should have minimal effect. I've done a few runs comparing 'before' and after' and found no difference at all. 
The only part I've split out into functions is the setup the initial loop for each of the load generation patterns. These only get called once per load generation run, so the impact is essentially nothing. 
Do you want me to perform some more benchmarking and come back with data?

gseddon commented:
Would it be better if I split this into smaller changes?

hatoo commented:
> Would it be better if I split this into smaller changes?

Sorry for delay.
Yeah it's very helpful if you split to smoller PRs.

hatoo commented:
Could you make this PR only contain `HttpWorkType` stuffs?
We can discuss about refactoring on other PRs. I think general idea of refactoring is good. Thank you.

hatoo commented:
>  is it a good thing for oha to have different performance depending on the way a user invokes it? For example, all of the qps paths will perform say 1% worse because they use the channels. So the user might see a different performance of their application depending on whether they use -q or not, or even perhaps between the -n and -z flags. So is it better to have the same function used for all of these paths so that the user gets consistent performance?

Yeah, performance variation depending on command line flags is not good.
But keeping codes optimal is more important to me.

> Also then if you performance optimise that one function, then all of the load tests automatically get faster.

I'm not saying I'll refuse all refactoring. I think we can refactoring work* to use common function while keeping the current performance by using generics.

hatoo commented:
Done at #746
